### PR TITLE
Compatibility with sparklyr

### PR DESCRIPTION
**Pull request by mgirlich**
Needed for sparklyr (https://github.com/sparklyr/sparklyr/pull/3245).

* Make `simulate_vars()` a generic: As tidyselect does not support lazy tables (https://github.com/r-lib/tidyselect/issues/242) we need `simulate_vars()`. In dbplyr we do not know the column types, in sparklyr the types are known. To not duplicate code and make selection with type predicates work the easiest solution seems to be to export `simulate_vars()` and make it generic.

### CODE REVIEW

File: R/verb-mutate.R [hunk 1]
```diff
@@ -162,7 +162,7 @@ get_mutate_layers <- function(.data, ...) {
     }
 
     all_vars <- names(cur_layer)[!var_is_null]
-    cur_data <- simulate_lazy_tbl(all_vars, grps)
+    cur_data$lazy_query <- add_select(cur_data, cur_layer, "mutate")
```
Comment (mgirlich) [inline R/verb-mutate.R:165]: This allows sparklyr to keep track of the column types

File: R/verb-select.R [hunk 2]
```diff
@@ -83,26 +83,54 @@ rename_with.tbl_lazy <- function(.data, .fn, .cols = everything(), ...) {
...
 }
 
-simulate_vars <- function(x, drop_groups = FALSE) {
+#' Simulate variables to use in tidyselect
+#'
+#' @param x A lazy table
+#' @param drop_groups Should groups be dropped?
+#'
+#' @return A 0 row tibble with the same columns as `x` and - if possible - the
+#'   column should be the R equivalend.
```
Comment (hadley) [inline R/verb-select.R:117]:
```suggestion
#' @return A 0 row tibble with the same columns names, and, if possible, types, as `x`.
```

File: R/verb-select.R [hunk 3]
```diff
@@ -83,30 +83,56 @@ rename_with.tbl_lazy <- function(.data, .fn, .cols = everything(), ...) {
 #' @inheritParams dplyr::relocate
 #' @export
 relocate.tbl_lazy <- function(.data, ..., .before = NULL, .after = NULL) {
+  sim_data <- simulate_vars(.data)
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/verb-select.R:89]: Why not just take the output from the new `simulate_vars()` and use it to re-make the previous data structure?

```R
sim_data <- simulate_vars(.data)
idx_data <- as_tibble(set_names(seq_along(sim_data), names(sim_data))
```

**Review by hadley — COMMENTED** [review]
We'll need some way to record in backends return correctly typed `sim_data()`, in order to correctly fix #583.

File: R/verb-select.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgirlich) [reply R/verb-select.R:89]: I don't think so due to relocating by column types e.g. `relocate(iris, where(is.character))`

[the remark this answers is not in the document]
Comment (hadley) [reply R/verb-select.R:89]: Oooo, gotcha.

File: R/verb-select.R [hunk 4]
```diff
@@ -83,26 +83,53 @@ rename_with.tbl_lazy <- function(.data, .fn, .cols = everything(), ...) {
...
+#' @return A 0 row tibble with the same columns names, and, if possible, types, as `x`.
+#'
+#' @export
+#' @keywords internal
+simulate_vars <- function (x, drop_groups = FALSE) {
+  UseMethod("simulate_vars")
+}
+
+#' @export
+simulate_vars.tbl_lazy <- function(x, drop_groups = FALSE) {
```
Comment (hadley) [inline R/verb-select.R:125]: I'd suggested adding something like:

```R
simulate_vars_is_typed <- function(x) UseMethod("simulate_vars_is_typed")
simulate_vars_is_typed.default <- function(x) FALSE
```

Then sparklyr can implement that generic at the same time, and we'll be in a good place for the next release of tidyselect. If you do that, I think it's fine to merge this PR.

**Pull request merged**